src/functions.rs

```diff
@@ -2827,23 +2911,109 @@ fn validate_argument<'a>(
+                Err(validate_argument_error_handler(
+                    arg,
+                    arg_typeinfo,
+                    target_argument,
+                    types,
                     file,
-                    arg.get_range(),
-                    &format!(
-                        "This type should inherit {}",
-                        target_argument.param_type.name
-                    ),
+                    func_info,
+                    func_map,
                 ))
             }
         }
     }
 }
 
+// Helper function for handling the error case for validate_argument.
+fn validate_argument_error_handler(
+    arg: ArgForValidation,
+    arg_typeinfo: &TypeInfo,
+    target_argument: &FunctionArgument,
+    types: &TypeMap,
+    file: Option<&SimpleFile<String, String>>,
+    func_info: Option<&FunctionInfo>,
+    func_map: Option<&FunctionMap>,
+) -> ErrorItem {
+    // If func_info is none, which means we are validating a built in and cannot be an associated call,
+    // or the call is not an asociated one fall through and return the "standard" error.
+    // If this is the associated call we need to do some more digging to give the user a better
+    // error message.
+    // Unwraps of func_info are safe in this block because of the false return on map_or
+    if func_info.map_or(false, |f| f.is_associated_call) {
+        let mut error = ErrorItem::make_compile_or_internal_error(
+            &format!(
+                "Expected type inheriting {} for associated call",
+                target_argument.param_type.name
+            ),
+            arg_typeinfo.get_file().as_ref(),
+            arg_typeinfo.name.get_range(),
+            &format!(
+                "An associated call: '{}' was made for this domain.  That call requires this domain inherit {}",
+                func_info.unwrap().name, target_argument.param_type.name
+            ),
+        );
+        if let (file, Some(range)) = (
+            func_info.unwrap().declaration_file,
+            func_info.unwrap().get_name_range(),
+        ) {
+            // We have a valid file name so we can put the associated function directly into the error
+            if !file.name().is_empty() {
+                if let ErrorItem::Compile(e) = error {
+                    error = ErrorItem::Compile(e.add_additional_message(
+                        file,
+                        range,
+                        "Associated function found here",
+                    ));
+                }
+                return error;
+            }
+        }
+
+        // The file name is not valid so we have to go looking through our parents to find out where
+        // we are called from.
+        if let Some(class) = types.get(
+            func_info
+                .unwrap()
+                .class
+                .get_name()
+                .unwrap_or(&CascadeString::from(""))
+                .as_ref(),
+        ) {
+            for parent in &class.inherits {
```

Comment:
Don't we need to recurse here?  What if the call was derived through a grandparent?

Reply:
What would the .cas look like for that?

I added some recursion code to try and fix it and I have something like:
```
virtual domain userdomain {}

virtual resource grandpa_user_tmp {
    @associated_call
	fn associated_call_from_user_tmp(userdomain source) {
		allow(source, user_tmp, file, [read]);
	}
}

virtual resource user_tmp inherits grandpa_user_tmp {}

@associate([user_tmp])
domain bad_userdomain {
	// Errors because bad_userdomain does not inherit userdomain, which
    // the associated call above requires
}
```

This compiles.  In this out.cil I don't see any macro `user_tmp-grandpa_user_tmp-associated_call_from_user_tmp`